ts/components/conversation/message/message-content/MessageReactBar.tsx

```diff
@@ -51,45 +51,132 @@ const ReactButton = styled.span`
   }
 `;
 
-export const MessageReactBar = (props: Props): ReactElement => {
-  const { action, additionalAction } = props;
-  const [recentReactions, setRecentReactions] = useState<RecentReactions>();
+const StyledContainer = styled.div<{ expirationTimestamp: number | null }>`
+  position: absolute;
+  top: ${props => (props.expirationTimestamp ? '-106px' : '-56px')};
+  display: flex;
+  flex-direction: column;
+  min-width: 0;
+  align-items: flex-start;
+  left: -1px;
+`;
+
+export const MessageReactBar = ({ action, additionalAction, messageId }: Props): ReactElement => {
+  const recentReactions = getRecentReactions();
+  const expirationTimestamp = useIsRenderedExpiresInItem(messageId);
 
-  useMount(() => {
-    const reactions = new RecentReactions(getRecentReactions());
-    if (reactions && !isEqual(reactions, recentReactions)) {
-      setRecentReactions(reactions);
-    }
-  });
+  return (
+    <StyledContainer expirationTimestamp={expirationTimestamp}>
+      <StyledMessageReactBar>
+        {recentReactions &&
+          recentReactions.map(emoji => (
+            <ReactButton
+              key={emoji}
+              role={'img'}
+              aria-label={
+                nativeEmojiData?.ariaLabels ? nativeEmojiData.ariaLabels[emoji] : undefined
+              }
+              onClick={() => {
+                action(emoji);
+              }}
+            >
+              {emoji}
+            </ReactButton>
+          ))}
+        <SessionIconButton
+          iconColor={'var(--emoji-reaction-bar-icon-color)'}
+          iconPadding={'8px'}
+          iconSize={'huge'}
+          iconType="plusThin"
+          backgroundColor={'var(--emoji-reaction-bar-icon-background-color)'}
+          borderRadius="300px"
+          onClick={additionalAction}
+        />
+      </StyledMessageReactBar>
+      <ExpiresInItem expirationTimestamp={expirationTimestamp} />
+    </StyledContainer>
+  );
+};
+
+function formatExpiry({ expirationTimestamp }: { expirationTimestamp: number }) {
+  const diffMs = expirationTimestamp - Date.now();
+  const diff = moment(diffMs).utc();
+
+  if (diffMs <= 0) {
+    return `0s`;
+  }
+
+  const prefix = 'Message will expire in';
+
+  if (diff.isBefore(moment.utc(0).add(1, 'minute'))) {
+    return `${prefix} ${diff.seconds()}s`;
+  }
+
+  if (diff.isBefore(moment.utc(0).add(1, 'hour'))) {
+    const extraUnit = diff.seconds() ? ` ${diff.seconds()}s` : '';
+    return `${prefix} ${diff.minutes()}m${extraUnit}`;
+  }
+
+  if (diff.isBefore(moment.utc(0).add(1, 'day'))) {
+    const extraUnit = diff.minutes() ? ` ${diff.minutes()}m` : '';
+    return `${prefix} ${diff.hours()}h${extraUnit}`;
+  }
+
+  if (diff.isBefore(moment.utc(0).add(7, 'day'))) {
```

Comment:
Minor thing but it might be nice to move all these hardcoded values (1, 7, 31, etc.) to `ts/session/constants.ts` under the duration object or something? Just for a single source of truth

Reply:
idk, I think we should see how we deal with the moderation options later, but because we "moment"s here, I think it makes sense to keep it here

Reply:
Resolved